Dockerfile

```diff
@@ -1,5 +1,61 @@
-FROM gcr.io/distroless/static
+# syntax=docker/dockerfile:1.1-experimental
 
-COPY bin/manager /usr/bin/
+# Copyright 2021 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
 
-ENTRYPOINT /usr/bin/manager
+# Build the manager binary
+# Run this with docker build --build-arg builder_image=<golang:x.y.z>
+ARG builder_image=golang:1.16.2
+FROM ${builder_image} as builder
+WORKDIR /workspace
+
+# Run this with docker build --build-arg goproxy=$(go env GOPROXY) to override the goproxy
+ARG goproxy=https://proxy.golang.org
+ENV GOPROXY=$goproxy
+
+# Copy the Go Modules manifests
+COPY go.mod go.mod
+COPY go.sum go.sum
+
+# Cache deps before building and copying source so that we don't need to re-download as much
+# and so that source changes don't invalidate our downloaded layer
+RUN --mount=type=cache,target=/go/pkg/mod \
+    go mod download
+
+# Copy the sources
+COPY ./ ./
+
+# Cache the go build into the the Go’s compiler cache folder so we take benefits of compiler caching across docker build calls
+RUN --mount=type=cache,target=/root/.cache/go-build \
+    --mount=type=cache,target=/go/pkg/mod \
```

Comment:
While this preserves the cache across the multi stage docker build process, it doesn't cache between docker builds.

For example, In our dev environment's we'd have to re-download go mod every time this multi state docker build runs.

Reply:
Actually, as inspired by CAPI Dockerfile, the caching is handled appropriately.

Here is the 1st build:
```
make docker-build
...
[+] Building 795.6s (19/19) FINISHED
 => [internal] load build definition from Dockerfile                                                                                                                                                                     0.0s
 => => transferring dockerfile: 2.19kB                                                                                                                                                                                   0.0s
 => [internal] load .dockerignore                                                                                                                                                                                        0.0s
 => => transferring context: 2B                                                                                                                                                                                          0.0s
 => resolve image config for docker.io/docker/dockerfile:1.1-experimental                                                                                                                                                0.0s
 => CACHED docker-image://docker.io/docker/dockerfile:1.1-experimental                                                                                                                                                   0.0s
 => [internal] load build definition from Dockerfile                                                                                                                                                                     0.0s
 => => transferring dockerfile: 2.19kB                                                                                                                                                                                   0.0s
 => [internal] load metadata for gcr.io/distroless/static:nonroot                                                                                                                                                        1.3s
 => [internal] load metadata for docker.io/library/golang:1.16.2                                                                                                                                                         0.0s
 => [stage-1 1/3] FROM gcr.io/distroless/static:nonroot@sha256:bca3c203cdb36f5914ab8568e4c25165643ea9b711b41a8a58b42c80a51ed609                                                                                          0.0s
 => => resolve gcr.io/distroless/static:nonroot@sha256:bca3c203cdb36f5914ab8568e4c25165643ea9b711b41a8a58b42c80a51ed609                                                                                                  0.0s
 => => sha256:bca3c203cdb36f5914ab8568e4c25165643ea9b711b41a8a58b42c80a51ed609 1.67kB / 1.67kB                                                                                                                           0.0s
 => => sha256:213a6d5205aa1421bd128b0396232a22fbb4eec4cbe510118f665398248f6d9a 426B / 426B                                                                                                                               0.0s
 => => sha256:bff4de2cb7e1dd0ed9797c6e33688f32f2ff0293ecee6fa069051081710bb61b 478B / 478B                                                                                                                               0.0s
 => [builder 1/8] FROM docker.io/library/golang:1.16.2                                                                                                                                                                   0.0s
 => [internal] load build context                                                                                                                                                                                        0.7s
 => => transferring context: 21.07MB                                                                                                                                                                                     0.7s
 => [builder 2/8] WORKDIR /workspace                                                                                                                                                                                     0.0s
 => [builder 3/8] COPY go.mod go.mod                                                                                                                                                                                     0.0s
 => [builder 4/8] COPY go.sum go.sum                                                                                                                                                                                     0.0s
 => [builder 5/8] RUN --mount=type=cache,target=/go/pkg/mod     go mod download                                                                                                                                        751.2s
 => [builder 6/8] COPY ./ ./                                                                                                                                                                                             0.1s
 => [builder 7/8] RUN --mount=type=cache,target=/root/.cache/go-build     --mount=type=cache,target=/go/pkg/mod     go build .                                                                                          18.5s
 => [builder 8/8] RUN --mount=type=cache,target=/root/.cache/go-build     --mount=type=cache,target=/go/pkg/mod     CGO_ENABLED=0 GOOS=linux GOARCH=amd64     go build -a -ldflags "${ldflags} -extldflags '-static'"   23.1s
 => [stage-1 2/3] COPY --from=builder /workspace/manager .                                                                                                                                                               0.1s
 => exporting to image                                                                                                                                                                                                   0.2s
 => => exporting layers                                                                                                                                                                                                  0.2s
 => => writing image sha256:734f443b21a43060fdd5ceee4573172c2fcada70fe0662d0d83933441a73923f                                                                                                                             0.0s
 => => naming to localhost:5000/capk-manager-amd64:dev                                                                                                                                                                   0.0s
...
```

Reply:
@davidvossel I think you are partially right too. Like if you change the go.mod or the go.sum file you will have to download all dependencies again, but otherwise the COPY command should see that the file did not change and just go on. I did not think of this possibility when we discussed it. It may be good enough.

Reply:
oh i see. this works because of the lines which COPY then RUN the go mod download _before_ the lines that COPY/RUN the build. so container image layering helps us here as long as go.mod/go.sum do not change.

This makes sense from a container build perspective.

However, what do we want to do with `make manager` make target now? It would be nice to be able to build using a containerized environment during the dev process in a way that doesn't generate a new container image every time.

Reply:
@davidvossel Seems we have reached the agreement on building the production image part with this PR.

Regarding the `make manager` part, we actually could enforce to use containerized environment during the dev process (some of my past projects are enforced). To satisfy the caching requirement, I am thinking of two approaches:
1). Launch a build-util container to build the manager binary, and cache the dependencies with volume. Like what you propose in #42 

2). Perform `make docker build with the Dockerfile proposed by this PR. Then create a temporary container (without running) and copy the manager binary out of the container. Remove the temporary container. The caching will be implemented by the image layers implicitly. And I can see this approach have one less script to maintain.

Reply:
@rmohr and @davidvossel, if you are okay with this Dockerfile for production image, could you approve the PR? My internal CI is blocked by this.

Reply:
> However, what do we want to do with `make manager` make target now? It would be nice to be able to build using a containerized environment during the dev process in a way that doesn't generate a new container image every time.

@davidvossel I agree that would be nice. Let's continue iterating on that in your PR.

> @rmohr and @davidvossel, if you are okay with this Dockerfile for production image, could you approve the PR? My internal CI is blocked by this.

Works for me. I am not sure if prow will accept my lgtm, since I am not in the org. Let's see.